linker/shared/src/main/scala/org/scalajs/linker/checker/IRChecker.scala

```diff
@@ -499,15 +499,16 @@ private final class IRChecker(unit: LinkingUnit, reporter: ErrorReporter) {
 
       case ArrayLength(array) =>
         typecheckExpr(array, env)
-        if (!array.tpe.isInstanceOf[ArrayType])
+        if (array.tpe != NullType && !array.tpe.isInstanceOf[ArrayType])
           reportError(i"Array type expected but ${array.tpe} found")
 
       case ArraySelect(array, index) =>
         typecheckExpect(index, env, IntType)
         typecheckExpr(array, env)
         array.tpe match {
+          case NullType => // will NPE, but allowed.
           case arrayType: ArrayType =>
-            if (tree.tpe != arrayElemType(arrayType))
+            if (!isSubtype(arrayElemType(arrayType), tree.tpe))
```

Comment:
I think this change is not valid. The type of the selection must be exactly the type of elements, because an `ArraySelect` can be the `lhs` of an `Assign`.

Reply:
Huh, I see, yeah.

The reason I got to this is that the following is not trivially correct under the current IR checker:

```scala
val ArraySelect(array, index) = tree
ArraySelect(transformExpr(array), transformExpr(index))(tree.tpe)
```

`transformExpr` could narrow the type of `array`, leading to ill-typed IR. IIUC this is a consequence of unsound covariance of array types.

Not sure what the best course of action is here. Maybe we'd have to allow `Assign(ArraySelect(...), ...)` with invalid types?

Reply:
Ah! That is indeed problematic. It does seem to be a consequence of the unsound covariance of arrays.

The [JVM spec for `aastore`](https://docs.oracle.com/javase/specs/jvms/se21/html/jvms-4.html#jvms-4.10.1.9.aastore), i.e., their equivalent of `Assign(ArraySelect(...), ...)` for reference array types, says:

> An `aastore` instruction is type safe iff one can validly pop types matching `Object`, `int`, and an array of `Object` off the incoming operand stack yielding the outgoing type state.

So for writes, apparently they allow to write any `Object` to any `Object[]`. For example, writing a `String` into a `List[]` is considered valid.

I guess it would make sense to do the same thing, then. We should check whether the backends can indeed deal with that (I think they can).

Reply:
> I guess it would make sense to do the same thing, then.

Agreed. I suspected that something like this happens on the JVM.

Reply:
Hmm... I'm not sure I understand this. For which IR nodes do we not allow this?

In general, it seems `typecheckExpect` in the IR checker checks a subtyping relation ship with the expected type.

In practice, this means that, for example, `TryFinally` with a wider type than its inner block is allowed (OTOH, it seems `checkApplyGeneric` check the apply node's type exactly).

Is `TryFinally` just the exception to align with `TryCatch`? (there it seems we need to allow it lest we require type widening trees between different branches, which IIUC we don't).

Reply:
We don't allow that for any node. We used to allow it for `This()` nodes, but eventually we addressed that in
https://github.com/scala-js/scala-js/commit/f0205cdba665d5cc28679291c7aa5de949435561

`TryFinally` is a bit weird. We don't need to check the expected type at all because its `tpe = block.tpe` at creation time.

Even then, for `TryCatch` for example, the result of the `tree: TryCatch` node must be a `tree.tpe`. If its `block.tpe` is different, a boxing operation will be emitted inside the `TryCatch` to adapt the result of `block` to the result of `TryCatch`. This is the same everywhere. When we *expect* an expression of type `T`, we first *emit* the expression as its own type `S` then insert an adaptation from `S` to `T`. This is only correct if the expression actually emits what it claims to emit, i.e., an `S`. Otherwise we may perform double-adaptation, or no adaptation where one was needed.

Reply:
TY for the explanation re adaptation, it is very helpful.

It makes me wonder whether we have followed this consistently, to minimize the nodes that can actually generate adaptations, but that's definitely out of scope here.